runtime/service.go

```diff
@@ -171,96 +182,60 @@ func NewService(shimCtx context.Context, id string, remotePublisher shim.Publish
 		}
 	}()
 
+	vmID := os.Getenv(internal.VMIDEnvVarKey)
+	logger := log.G(shimCtx)
+	if vmID != "" {
+		logger = logger.WithField("vmID", vmID)
+	}
+
 	s := &service{
 		taskManager: vm.NewTaskManager(log.G(shimCtx)),
 
-		server:        server,
 		publish:       remotePublisher,
 		eventExchange: eventExchange,
 		namespace:     namespace,
 
-		logger:     log.G(shimCtx),
+		logger:     logger,
 		shimCtx:    shimCtx,
 		shimCancel: shimCancel,
 
-		vmID:   os.Getenv(vmIDEnvVarKey),
+		vmID:   vmID,
 		config: config,
-	}
-
-	return s, nil
-}
-
-// vmDir holds files, sockets and FIFOs scoped to a single given VM.
-// It is unique per-VM and containerd namespace.
-func (s *service) vmDir() vm.Dir {
-	return vm.Dir(filepath.Join(varRunDir, s.namespace, s.vmID))
-}
-
-func (s *service) newShim(namespace string, containerdBinary, containerdAddress string, shimSocket *net.UnixListener) (*exec.Cmd, error) {
-	self, err := os.Executable()
-	if err != nil {
-		return nil, err
-	}
-
-	args := []string{
-		"-namespace", namespace,
-		"-address", containerdAddress,
-		"-publish-binary", containerdBinary,
-	}
-
-	if s.config.Debug {
-		args = append(args, "-debug")
-	}
 
-	cmd := exec.Command(self, args...)
-
-	cmd.Dir = s.vmDir().RootPath()
-
-	cmd.Env = append(os.Environ(),
-		fmt.Sprintf("%s=%s", vmIDEnvVarKey, s.vmID))
-
-	shimSocketFile, err := shimSocket.File()
-	if err != nil {
-		return nil, err
-	}
-	cmd.ExtraFiles = append(cmd.ExtraFiles, shimSocketFile)
-
-	cmd.SysProcAttr = &syscall.SysProcAttr{
-		Setpgid: true,
+		vmReady: make(chan struct{}),
 	}
 
-	err = cmd.Start()
-	if err != nil {
-		return nil, err
-	}
+	// TODO we have to create separate listeners for the fccontrol service and shim service because
+	// containerd does not currently expose the shim server for us to register the fccontrol service with too.
+	// This is likely addressable through some relatively small upstream contributions; the following is a stop-gap
+	// solution until that time.
+	fcSocketFDEnvVal := os.Getenv(internal.FCSocketFDEnvKey)
```

Comment:
Where is `FCSocketFDEnvKey` setting the file descriptor?

Reply:
local.go sets the env var and the FD in `newShim` (the FD is set as part of `ExtraFiles`)